File: stack/validParentheses.py


class Solution:
    @staticmethod
    def isValid(s: str) -> bool:
        stack = []
        closeToOpen = {")": "(", "}": "{", "]": "["}
        for c in s:
            if c in closeToOpen:
                if stack and stack[-1] == closeToOpen[c]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(c)
        return not stack


if __name__ == "__main__":
    s = "([{}])"
    print(Solution.isValid(s))

File: stack/cppRealization/validParentheses.cpp
#include <iostream>
#include <stack>
#include <map>

using namespace std;


class Solustion {
public:
    static bool isValid(string& s){
        stack<char> stack;
        map<char, char> closeToOpen = {{')', '('}, {']', '['}, {'}', '{'}};
        for(char c : s){
            if(closeToOpen.find(c) != closeToOpen.end()){
                if(!stack.empty() && stack.top() == closeToOpen[c]){
                    stack.pop();
                }else{
                    return false;
                }
            }else{
                stack.push(c);
            }
        }
        return stack.empty();
    }
};


int main(){
    string s = "([{}])";
    if(Solustion::isValid(s) == true){
        cout << "True" << endl;
    }else{
        cout << "False" << endl;
    }
    return 0;
}